### ISSUE TITLE
Don't create group activity stream items when doc read settings are restrictive.

### ISSUE THREAD

**Issue by dcavins** [boonebgorges/buddypress-docs#532]
If a doc is associated with a group but only readable by group mods or admins or the creator (which is in itself weird, but done) we shouldn't create an activity item because it's a tease that results in 404s.

Here's one approach: https://github.com/dcavins/buddypress-docs/commit/9905ced0175fc51f4a6ef9c86411f25d894370fa

I have reservations about this because it moves the `bp_docs_activity_args` filter to fire before the `bp_docs_before_activity_save` action, would could be surprising, because the group info is added at `bp_docs_activity_args`. Also, destroying the activity args to halt the process is effective, but could cause problems with filters that come later (and expect `$args['secondary_item_id']` to exist).

Another thought was to add a boolean filter `bp_docs_allow_activity_item_creation` before `bp_docs_before_activity_save`, but this would still require the activity args filter to be moved up in the chain. And it adds another filter. But maybe another filter that does what it says is more straightforward:
https://github.com/dcavins/buddypress-docs/commit/64a0c9d3d9778fc712b66ae5224241f9d4a16fd6

Comment activity is directly hot-wired, already: https://github.com/boonebgorges/buddypress-docs/blob/master/includes/activity.php#L43
So we could also go that route. (Or the comment code could use the new `bp_docs_allow_activity_item_creation` filter if we go that way. )

**Comment by boonebgorges** [comment]
Short circuiting the creation of activity items as in dcavins@9905ced seems like it'd be frustrating for developers.

My preference, all things being equal, is to be maximally permissive about which activity items get created, and then to filter them out on display. In this case, that would mean creating the activity items, but making sure they're filtered out in activity streams for users who shouldn't be able to see them. Here's an example from another project I recently worked on: https://github.com/cuny-academic-commons/social-paper/blob/master/includes/hooks-buddypress-activity.php#L444 

One downside of this approach is that certain plugins (like buddypress-group-email-subscription) will perform actions at the creation of group activity items, whether or not they're filtered on display, and so need to be specifically handled in order to prevent bogus notifications from going out. I'm not sure if there's a way around this, without extensive refactoring within BP itself.

If you can't see a way to make the create-then-filter approach work, something like dcavins@64a0c9d seems more straightforward (though still kinda homely).

**Comment by dcavins** [comment]
Here's some work that addresses problems with the visibility of docs activity items in activity streams and comment visibility in activity streams and via `get_comments()`, so applies to this issue as well as #504.
This is the first working pass at this. I'm sure that the changes to Docs_Access_Query can be streamlined, but this is the "get it working" pass.
https://github.com/dcavins/buddypress-docs/commits/activity-stream-exclusions

**Comment by boonebgorges** [comment]
Looks like a good approach. A couple questions:

- I couldn't tell at a glance whether including the comment-protection 
query will introduce overhead to Doc queries that don't involve 
comments. Normal `get_posts()` calls shouldn't need to fetch any info 
about comment access, right? Can you verify that no database calls are 
being added in these cases?
- This doesn't touch activity yet, right?
- Unit tests? :)

**Comment by dcavins** [comment]
Thanks for your thoughtful feedback.

Yes, I watched the additional queries very closely as I worked, because this is something that I worry about a lot on my installation.

Here's what I recorded along the way, using Query Monitor, and logged in as a non-site-admin so that filtering would kick in:
/activity, without doc-related item filtering
2.56S   25.74MB   0.1454S   276Q
with filtering, no singleton storage
4.86S   28.78MB   1.0780S   319Q
with filtering and singleton storage:
2.68S   26.56MB   0.1973S   277Q

I'm not seeing the comment-array-building WP_Query running when comments aren't involved. I'll keep looking, though. Using the singleton storage means that those heavy queries run once each per page load, which is sure a lot better.

Activity: per-item protection for create/edit and comments is introduced here: 
https://github.com/dcavins/buddypress-docs/commit/93082178023cda0858ed89da8874b5edbdf42bb1#diff-a3a6add1690a16cb3830892d6e44306cR433

Unit tests: I agree entirely. We've got to create a `save_doc()` function that doesn't rely on $_POST so we can get serious about unit tests for the entire lifespan of a doc. I worked on that once, and will refresh this: https://github.com/dcavins/buddypress-docs/commits/restructure_save_2 . (The is needed to test docs-related activity creation, among other things.) :)

Optimization opportunities: `wp_get_object_terms()` runs the same queries again and again. You recently made some optimizations (using `get_terms()` i think), we could do more. BP needs to cache group membership status. We check that over and over again.

Thanks again!

**Comment by boonebgorges** [comment]
> I'm not seeing the comment-array-building WP_Query running when comments aren't involved.

Cool. I will keep a look-out too. Again, my question was based on a very brief study :-D

> Activity: per-item protection for create/edit and comments is introduced here: 

Whoops, missed that.

> We've got to create a save_doc() function that doesn't rely on $_POST so we can get serious about unit tests for the entire lifespan of a doc. I worked on that once, and will refresh this: https://github.com/dcavins/buddypress-docs/commits/restructure_save_2 . (The is needed to test docs-related activity creation, among other things.) :)

Make it so.

> You recently made some optimizations (using get_terms() i think),

`get_the_terms()`, but yeah. We should  switch to that where possible.

> BP needs to cache group membership status. 

Big time. I think I wrote a patch for that, but we tabled it because of the work you were doing on invitation stuff.

**Comment by dcavins** [comment]
I've moved the singleton storage to wp_cache and added tests that verify that the cache is preventing multiple database dips in this change set:
https://github.com/dcavins/buddypress-docs/commit/abde079d049c63eb13ede31c4469583e82ac4105

I'll work on general tests for the access protection functions.

**Comment by dcavins** [comment]
I had a few minutes this afternoon, so thought I could fill in some tests for `$bp_docs_access_query->get_doc_ids()`, `$bp_docs_access_query->get_restricted_comment_doc_ids()`, and `$bp_docs_access_query->get_comment_ids()` but I'm having a problem. When `get_doc_ids()` is run as part of a page load, in the `forbidden_fruit` `WP_Query` I get the right `tax_query` and it's incorporated into the `request`:
```
["tax_query"]=>
  object(WP_Tax_Query)#424 (6) {
    ["queries"]=>
    array(1) {
      [0]=>
      array(5) {
        ["taxonomy"]=>
        string(14) "bp_docs_access"
        ["terms"]=>
        array(1) {
          [0]=>
          string(21) "bp_docs_access_anyone"
        }
        ["field"]=>
        string(4) "slug"
        ["operator"]=>
        string(6) "NOT IN"
        ["include_children"]=>
        bool(true)
      }
    }
    ["relation"]=>
    string(3) "AND"
    ["table_aliases":protected]=>
    array(0) {
    }
    ["queried_terms"]=>
    array(0) {
    }
    ["primary_table"]=>
    string(8) "wp_posts"
    ["primary_id_column"]=>
    string(2) "ID"
  }
["request"]=>
  string(297) "SELECT   wp_posts.ID FROM wp_posts  WHERE 1=1  AND ( 
  wp_posts.ID NOT IN (
				SELECT object_id
				FROM wp_term_relationships
				WHERE term_taxonomy_id IN (2)
			)
) AND wp_posts.post_type = 'bp_doc' AND (wp_posts.post_status = 'publish') 
GROUP BY wp_posts.ID ORDER BY wp_posts.post_date DESC "
```

But, when running tests, the `request` doesn't seem to take the `tax_query` into account (and seems to not find any docs, even though the request as written should return every doc, I think).
```
["tax_query"]=>
  object(WP_Tax_Query)#513 (6) {
    ["queries"]=>
    array(1) {
      [0]=>
      array(5) {
        ["taxonomy"]=>
        string(14) "bp_docs_access"
        ["terms"]=>
        array(1) {
          [0]=>
          string(21) "bp_docs_access_anyone"
        }
        ["field"]=>
        string(4) "slug"
        ["operator"]=>
        string(6) "NOT IN"
        ["include_children"]=>
        bool(true)
      }
    }
    ["relation"]=>
    string(3) "AND"
    ["table_aliases":protected]=>
    array(0) {
    }
    ["queried_terms"]=>
    array(0) {
    }
    ["primary_table"]=>
    string(13) "wptests_posts"
    ["primary_id_column"]=>
    string(2) "ID"
  }
["request"]=>
  string(204) "SELECT   wptests_posts.ID FROM wptests_posts  WHERE 1=1  
  AND wptests_posts.post_type = 'bp_doc' AND (wptests_posts.post_status = 'publish') 
  GROUP BY wptests_posts.ID ORDER BY wptests_posts.post_date DESC "
```

Here's an example test checking that a doc is off-limits to a specific user:
```php
	public function test_bp_docs_access_query_get_doc_ids() {
		$old_current_user = get_current_user_id();

		$d = $this->factory->doc->create();
		bp_docs_update_doc_access( $d, 'loggedin' );

		// Pretend we're logged out--this user shouldn't have access.
		$this->set_current_user( 0 );

		$bp_docs_access_query = bp_docs_access_query();
		$restricted_ids = $bp_docs_access_query->get_doc_ids();

		$this->assertTrue( in_array( $d, $restricted_ids ) );

		$this->set_current_user( $old_current_user );
	}
```
Any ideas about what I'm doing wrong?

**Comment by dcavins** [comment]
It looks like BuddyPress Group Email Subscription needs to add a filter so that we can prevent items from being added to the digest and summary emails. If there were a filter in `ass_digest_record_activity()` we could catch those. We can currently stop the immediate emails using the `bp_ass_send_activity_notification_for_user` filter.

I was hoping we could use `ass_block_group_activity_types` to stop the process, but it happens before the recipient loop, so isn't recipient aware. (It only has access to the activity item object.)

I'll submit a PR against BPGES with a filter.

**Comment by boonebgorges** [comment]
I just merged https://github.com/boonebgorges/buddypress-group-email-subscription/pull/84. I'll have a look at your tax_query question now.

**Comment by boonebgorges** [comment]
Hi @dcavins - I'm not 100% sure what's going on above. Miscellaneous thoughts:

- The test you've supplied is passing for me. Is that what you're seeing?
- `request` in your first var_dump doesn't contain a tax clause at all. This suggests to me that the `bp_docs_access` taxonomy isn't being registered in time, or something like that. Follow the `WP_Tax_Query` chain to see where it's bailing.
- `SELECT   wptests_posts.ID FROM wptests_posts  WHERE 1=1  AND wptests_posts.post_type = 'bp_doc' AND (wptests_posts.post_status = 'publish') GROUP BY wptests_posts.ID ORDER BY wptests_posts.post_date DESC` ought to return all published `bp_doc` posts. Make sure that `post_status=publish` on your posts in the db. If it's returning nothing at all, and you're certain the DB is set up correctly, then something very weird must be happening - check to see if there are any filters directly on sql queries (such as `add_filter( 'query' ... )`)

**Comment by dcavins** [comment]
I think this is the source of the problem:
https://github.com/dcavins/buddypress-docs/commit/a2cea134ef17890e1946312ea5076ea27caa0bec

**Comment by dcavins** [comment]
This will be resolved in BP 2.0. The fix has been committed to the 2.0 branch, here is a key commit:
9308217
